controllers/scopeinstance_controller.go

```diff
@@ -104,210 +131,167 @@ func (r *ScopeInstanceReconciler) Reconcile(ctx context.Context, req ctrl.Reques
-		scopeTemplateUIDKey: string(st.GetUID()),
-	}
+	return nil
+}
 
-	requirement, err = labels.NewRequirement(scopeTemplateHashKey, selection.NotEquals, []string{util.HashObject(st.Spec)})
-	if err != nil {
-		return ctrl.Result{}, err
+func (r *ScopeInstanceReconciler) createOrUpdateClusterRoleBinding(ctx context.Context, cr *operatorsv1.ClusterRoleTemplate, in *operatorsv1.ScopeInstance, st *operatorsv1.ScopeTemplate) error {
+	crb := r.getClusterRoleBinding(cr, in, st)
+	crbList := &rbacv1.ClusterRoleBindingList{}
+	if err := r.Client.List(ctx, crbList, client.MatchingLabels{
+		scopeInstanceUIDKey:           string(in.GetUID()),
+		clusterRoleBindingGenerateKey: cr.GenerateName,
+	}); err != nil {
+		return err
 	}
 
-	listOptions = &client.ListOptions{
-		LabelSelector: labels.NewSelector().Add(*requirement),
+	if len(crbList.Items) > 1 {
+		return fmt.Errorf("more than one ClusterRoleBinding found for ClusterRole %s", cr.GenerateName)
 	}
 
-	if err := r.deleteBindings(ctx, listOption, listOptions); err != nil {
-		log.Log.Error(err, "in deleting Role Bindings")
-		return ctrl.Result{}, err
+	// GenerateName is immutable, so create the object if it has changed
+	if len(crbList.Items) == 0 {
+		if err := r.Client.Create(ctx, crb); err != nil {
+			return err
+		}
+		return nil
 	}
 
-	log.Log.Info("No ScopeInstance error")
-
-	return ctrl.Result{}, nil
-}
+	existingCRB := &crbList.Items[0]
+	if util.IsOwnedByLabel(existingCRB.DeepCopy(), in) &&
+		reflect.DeepEqual(existingCRB.Subjects, crb.Subjects) &&
+		reflect.DeepEqual(existingCRB.Labels, crb.Labels) {
+		log.Log.Info("existing ClusterRoleBinding does not need to be updated")
+		return nil
+	}
 
-func (r *ScopeInstanceReconciler) ensureBindings(ctx context.Context, in *operatorsv1.ScopeInstance, st *operatorsv1.ScopeTemplate) error {
-	// it will create clusterrole as shown below if no namespace is provided
-	// TODO: refactor code to handle both roleBindings and clusterRoleBindings
-	if len(in.Spec.Namespaces) == 0 {
-		for _, cr := range st.Spec.ClusterRoles {
-			crb := &rbacv1.ClusterRoleBinding{
-				ObjectMeta: metav1.ObjectMeta{
-					GenerateName: cr.GenerateName + "-",
-					Labels: map[string]string{
-						scopeInstanceUIDKey:           string(in.GetUID()),
-						scopeTemplateUIDKey:           string(st.GetUID()),
-						scopeInstanceHashKey:          util.HashObject(in.Spec),
-						scopeTemplateHashKey:          util.HashObject(st.Spec),
-						clusterRoleBindingGenerateKey: cr.GenerateName,
-					},
-					OwnerReferences: []metav1.OwnerReference{{
-						APIVersion: in.APIVersion,
-						Kind:       in.Kind,
-						Name:       in.GetObjectMeta().GetName(),
-						UID:        in.GetObjectMeta().GetUID(),
-					}},
-				},
-				Subjects: cr.Subjects,
-				RoleRef: rbacv1.RoleRef{
-					Kind:     "ClusterRole",
-					Name:     cr.GenerateName,
-					APIGroup: "rbac.authorization.k8s.io",
-				},
-			}
+	u := r.patchConfigForClusterRoleBinding(existingCRB, crb)
```

Comment:
Why did you make this its own method? You're just creating an object to pass to the `patchBinding` method. I would either just create the object inline at the risk of creating more noise in the method. OR rename the method to something like `wrapClusterRoleInUnstructered` or `prepareForPatching` something that shows what you are actually doing. Because right now it looks like you are patching a config on a CRB AND then patching the binding which is not what you really are doing.